lib/cdo/contact_rollups/v2/pardot.rb

```diff
@@ -24,40 +24,51 @@ def initialize
 
   # Retrieves new (email, Pardot ID) mappings from Pardot
   #
-  # @yieldreturn [Array<Hash>] an array of hash {email, id}
-  # @raise [StandardError] if receives errors in Pardot response
-  #
   # @param [Integer] last_id retrieves only Pardot ID greater than this value
+  # @param [Array<String>] a list of fields. Must have 'id' the list. Field names must be url-safe.
+  # @param [Integer] limit maximum number of prospects to retrieve
   # @return [Integer] number of results retrieved
-  def self.retrieve_new_ids(last_id)
+  #
+  # @yieldreturn [Array<Hash>] an array of hash of prospect data
+  # @raise [ArgumentError] if 'id' is not in the list of fields
+  # @raise [StandardError] if receives errors in Pardot response
+  def self.retrieve_prospects(last_id, fields, limit = nil)
+    raise ArgumentError.new("Missing value 'id' in fields argument") unless fields.include? 'id'
     total_results_retrieved = 0
 
     # Run repeated requests querying for prospects above our highest known Pardot ID.
     loop do
-      url = "#{PROSPECT_QUERY_URL}?id_greater_than=#{last_id}&fields=email,id&sort_by=id"
+      url = "#{PROSPECT_QUERY_URL}?id_greater_than=#{last_id}&fields=#{fields.join(',')}&sort_by=id"
+      url += "&limit=#{limit}" if limit
       doc = post_with_auth_retry(url)
       raise_if_response_error(doc)
 
       # Pardot returns the count total available prospects (not capped to 200),
       # although the data for a max of 200 are contained in the response.
+      # @see http://developer.pardot.com/kb/api-version-4/prospects/#xml-response-format
       total_results = doc.xpath('/rsp/result/total_results').text.to_i
 
       results_in_response = 0
-      mappings = []
+      prospects = []
       doc.xpath('/rsp/result/prospect').each do |node|
-        id = node.xpath("id").text.to_i
-        email = node.xpath("email").text
-        mappings << {email: email, pardot_id: id}
-        last_id = id
+        prospect = {}
+        fields.each do |field|
+          value = node.xpath(field).text
+          prospect.merge!({field => value})
+        end
+        prospects << prospect
+
+        last_id = prospect['id']
         results_in_response += 1
       end
 
-      yield mappings if block_given?
+      yield prospects if block_given?
       log "Retrieved #{results_in_response}/#{total_results} new Pardot IDs. Last Pardot ID = #{last_id}."
 
-      # Stop if all the remaining results were in this response
+      # Stop if all the remaining results were in this response or we reached the download limit.
       total_results_retrieved += results_in_response
-      break if results_in_response == total_results
+      break if results_in_response == total_results ||
+        (limit && total_results_retrieved >= limit)
```

Comment:
I still think there's a possible infinite loop here.

1. Suppose `retrieve_prospects` is called with `limit = nil` and there are 300 total results to retrieve.
2. On our first loop we retrieve 200 prospects. At line 66 we log "Retrieved 200/300 new Pardot IDs"
3. On our second loop, we reset `results_in_response` to 0 on line 51, and log "Retrieved 100/300 new Pardot IDs" at line 66.
4. We should break at line 70, but we do not, because `results_in_response` (100) does not equal `total_results` (300) and `limit` is nil.

I think the solution here is to use `total_results_retrieved` instead of `results_in_response` on line 70.

Reply:
I think that's what I was trying to say in my previous comment -- I think `total_results` gets reduced to 100 the second time through the loop?

Reply:
I see the confusion. Ben is right, `total_results` is updated for each request
https://github.com/code-dot-org/code-dot-org/blob/080db54723f5b5af521a230d61d2ecdc30b412ba/lib/cdo/contact_rollups/v2/pardot.rb#L49
because the `url` we sent to Pardot is updated with new `id_greater_than` value.
https://github.com/code-dot-org/code-dot-org/blob/080db54723f5b5af521a230d61d2ecdc30b412ba/lib/cdo/contact_rollups/v2/pardot.rb#L41
So at step 3 in Brad's list, it will log "Retrieved 100/10 new Pardot IDs".
I'll add a comment to clarify this behavior.

Reply:
Oh! I understand now.  That is what confused me, I was expecting `total_results` to remain constant.

Reply:
That seems like a reasonable assumption :)